src/wwwroot/js/genpage/main.js

```diff
@@ -700,9 +700,25 @@ function setCurrentImage(src, metadata = '', batchId = '', previewGrow = false,
                 'width': width * 2,
                 'height': height * 2
             };
-            mainGenHandler.doGenerate(input_overrides, { 'initimagecreativity': 0.4 });
+            let upscaleCreativity = parseFloat(getUserSetting('UpscaleCreativity', '0.4'));
+            mainGenHandler.doGenerate(input_overrides, { 'initimagecreativity': upscaleCreativity });
         }));
     }, '', 'Runs an instant generation with this image as the input and scale doubled');
+    includeButton('Refine Image', () => {
+        toDataURL(img.src, (url => {
+            let input_overrides = {
+                'images': 1,
+            };
+            let toggler = getRequiredElementById('input_group_content_refineupscale_toggle');
+            toggler.checked = true;
+            triggerChangeFor(toggler);
+            reuseLastParamVal('input_seed');
```

Comment:
This reuse/setrandom is not great here, use input overrides

Reply:
... also this code in general doesn't do what you'd expect a refine button to do. Nowhere here are you inputting the relevant image, for example.

Reply:
> ... also this code in general doesn't do what you'd expect a refine button to do. Nowhere here are you inputting the relevant image, for example.

I'm not sure what you mean, this does exactly what I expect it to do with the current image.

Reply:
Generate a picture of a cat, change the prompt box to say "waffle fries", then click the Refine button. What happens? Does it refine the image, or does it do something else entirely?

Reply:
Alternately, drag a random image that didn't come from an AI at all into Swarm's interface, and click Refine. What does it do? Does it refine that image, or something else entirely?

Reply:
Ok, I see what you mean. It doesn't work as intended when it's not used as intended. I'll have to fix that then

Reply:
lol, if the intent of a button *on an image* that says "Refine" isn't to refine the image, you've done something very wrong somewhere in planning

Reply:
No, that's exactly the intention, and that's what it does. Changing the prompt or the image itself before clicking the refine button is not the intended use, although it still does what it says it does. What I had not considered in my planning was that it would be used in that way.

The whole point is to use the current refine settings on the current image. It does that perfectly. What it does not do, is account for discrepancies between the current image and other settings since I didn't plan for someone to change anything before pressing the button.

Reply:
You wrote code that adds a button every time any image is displayed in the center area of Swarm's interface, but only does anything of use if the image is fresh out of the generate button oven. The placement and your stated intention do not align. Technically you could get closer to your stated intention by moving the button somewhere else, but it's better to just allow refining images that aren't the most recently generated image.

Reply:
My use case is exactly as you say.

Generate an image at a low resolution, like the image and want to upscale the image, click the 'Refine Image' button and get an upscaled version using the current refine settings.

It's intended to work exactly as the 'Upscale 2x' button, but using different parameters since the 'Upscale 2x' button does not produce the results I want, but I like the easy convenience of a single click button.

I'll admit my hacky fix might not be the best way to accomplish it, but it does exactly as it was intended, in the use case it was intended for.

I'm not sure where else I would put the button since in my use case I have the 'Refine Image' button shown by default in the center area, and the 'Upscale 2x' button falls under the 'more' list because I don't use it. 

Unfortunately I'm not familiar enough with the code at the moment to make it use the actually displayed current image and it's existing parameters and adding the refine parameters to that. However, if that's the solution then that's what I'll work on and figure out.

Reply:
basically just feed the image as an init image, but override creativity to 0, and it oughtta work exactly how you're intending